// prepack/lib/react/errors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReconcilerFatalError = exports.DoNotOptimize = exports.NewComponentTreeBranch = exports.UnsupportedSideEffect = exports.SimpleClassBailOut = exports.ExpectedBailOut = undefined;

require("../serializer/types.js");

var _errors = require("../errors.js");

// ExpectedBailOut is like an error, that gets thrown during the reconcilation phase
// allowing the reconcilation to continue on other branches of the tree, the message
// given to ExpectedBailOut will be assigned to the value.$BailOutReason property and serialized
// as a comment in the output source to give the user hints as to what they need to do
// to fix the bail-out case
/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

/*  strict-local */

class ExpectedBailOut extends Error {}

exports.ExpectedBailOut = ExpectedBailOut; // SimpleClassBailOuts only occur when a simple class instance is created and used
// bailing out here will result in a complex class instance being created after
// and an alternative complex class component route being used

class SimpleClassBailOut extends Error {}

exports.SimpleClassBailOut = SimpleClassBailOut; // When the reconciler detectes a side-effect in pure evaluation, it throws one
// of these errors. This will fall straight through the the wrapping React
// component render try/catch, which will then throw an appropiate
// ReconcilerFatalError along with information on the React component stack

class UnsupportedSideEffect extends Error {}

exports.UnsupportedSideEffect = UnsupportedSideEffect; // NewComponentTreeBranch only occur when a complex class is found in a
// component tree and the reconciler can no longer fold the component of that branch

class NewComponentTreeBranch extends Error {
  constructor(evaluatedNode) {
    super();
    this.evaluatedNode = evaluatedNode;
  }
}

exports.NewComponentTreeBranch = NewComponentTreeBranch;
class DoNotOptimize extends Error {}

exports.DoNotOptimize = DoNotOptimize; // Used when an entire React component tree has failed to optimize
// this means there is a programming bug in the application that is
// being Prepacked

class ReconcilerFatalError extends _errors.FatalError {
  constructor(message, evaluatedNode) {
    super(message);
    evaluatedNode.status = "FATAL";
    evaluatedNode.message = message;
    this.evaluatedNode = evaluatedNode;
    // used for assertions in tests
    this.__isReconcilerFatalError = true;
  }
}
exports.ReconcilerFatalError = ReconcilerFatalError;
